### PR TITLE
Feature/config map

### PR DESCRIPTION
**Pull request by caaespin**
This pull request brings all the work done for the HCA demo back into a generalized, configurable fashion. Among other things, it:
- Maps the entries in the elasticsearch entries to the API format
- Use config file to define which facets to show
- Refactoring of the code into a couple baseclases
- Updating the flask endpoints
- Use of logger to log webservice related stuff
- Travis CI 
- A unit test suite, with scripts and makefile to automate testing
-PEP8 compliant 

That's it in a nutshell
Please help me by doing a code review. It is a very big pull request

Fixes BD2KGenomics/dcc-ops#118,  Fixes BD2KGenomics/dcc-ops#119

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -5,143 +5,13 @@ There are currently five working endpoints:
...
-```
-#Make the redwood-client directory and cd into the directory
-mkdir redwood-client
-cd redwood-client
-#And then just download the client
-wget https://s3-us-west-2.amazonaws.com/beni-dcc-storage-dev/20161216_ucsc-storage-client.tar.gz && tar zxf 20161216_ucsc-storage-client.tar.gz && rm -f 20161216_ucsc-storage-client.tar.gz
-```
-To set up the dashboard, follow the instructions https://github.com/BD2KGenomics/dcc-dashboard
-
-## Installing Apache
```
Comment (mjkrause) [inline README.md]: You had mentioned to me that we would not need Apache anymore, now that we use NGINX. Is it still needed for some other functionality?

Comment (caaespin) [reply README.md]: Let me update the README

File: README.md [hunk 2]
```diff
@@ -5,143 +5,13 @@ There are currently five working endpoints:
...
-To deploy the Flask application, we will use Apache. To install Apache and the required components in your machine, run the following commands:
-```
-# Install the apache webserver and mod_wsgi.
-sudo apt-get update
-sudo apt-get install apache2
-sudo apt-get install libapache2-mod-wsgi
-sudo a2enmod headers
-sudo service apache2 restart
-```
-## Make the Soft Link
```
Comment (mjkrause) [inline README.md]: "symbolic link" instead?

File: README.md [hunk 3]
```diff
@@ -5,143 +5,13 @@ There are currently five working endpoints:
...
-To enable the instructions.sh script, which contains the commands to run and update the Elasticsearch indicies that the web service uses, do:
-```
-chmod +x dcc-dashboard-service/instructions.sh
-```
-## Setting the 'instructions.sh' as a cron job
-To start the cron job, simply do:
-```
-crontab dcc-dashboard-service/crontab.txt
-```
-<b>NOTE:</b> crontab.txt assumes you have an environmental variable called `$REDWOOD_ACCESS_TOKEN`. To set up that variable, edit your bash file (in my case it was .bashrc). At the end of it, add
```
Comment (mjkrause) [inline README.md]: add a script that checks for the existence of $REDWOOD_ACCESS_TOKEN? found some ideas here: https://stackoverflow.com/questions/39296472/shell-script-how-to-check-if-an-environment-variable-exists-and-get-its-value

File: billing.py [hunk 4]
```diff
@@ -30,24 +22,28 @@
 es_service = os.environ.get("ES_SERVICE", "localhost")
 es = Elasticsearch(['http://' + es_service + ':9200/'])
 
+
 @login_manager.user_loader
 def load_user(user_id):
     return User.query.get(int(user_id))
```
Comment (mjkrause) [inline billing.py]: why using 3 lines between def's? is because of the decorator?

Comment (caaespin) [reply billing.py]: Yes. PEP8 will complain if there are no 2 blank lines between defs. If there is a decorator, the decorator needs to be right above the def, but two lines of space between the previous def

File: es_filebrowser_index.py [hunk 5]
```diff
@@ -1,225 +0,0 @@
-#indexer_es_fb.py
```
Comment (mjkrause) [inline es_filebrowser_index.py]: what happened to `es_filebrowser_index.py`? what is the reason you took it out?

Comment (caaespin) [reply es_filebrowser_index.py]: It got moved to a repo called `dcc-metadata-indexer`. It was used to generate the file oriented index. It was initially put here but then it made sense to concentrate all indexing code into the `dcc-metadata-indexer` repo.

File: mapi.py [hunk 6]
```diff
@@ -1,91 +1,54 @@
-from flask import Flask, jsonify, request, session, Blueprint
```
Comment (mjkrause) [inline mapi.py]: Hi Carlos, is `session` a module our group wrote? I can't seem to find a module with this name on Google. `Session` is a method in `request`, but that's not what is imported here. Just wondering.

File: billing.py [hunk 7]
```diff
@@ -1,27 +1,19 @@
...
-
 import os
-from models import Billing, db
-from utility import get_compute_costs, get_storage_costs, create_analysis_costs_json, create_storage_costs_json
+from models import Billing
+from utility import get_compute_costs, get_storage_costs,\
+    create_analysis_costs_json, create_storage_costs_json
 import datetime
 import calendar
 import click
```
Comment (mjkrause) [inline billing.py]: Hi Carlos, unrelated to this review. Are people using more `click` or more `argparse` in our group? What do you prefer?

File: mapi.py (hunk 6 above)
Comment (caaespin) [reply mapi.py]: `session` is a submodule within `flask`. We didn't write it. See: http://flask.pocoo.org/docs/0.12/api/#flask.session

File: billing.py (hunk 7 above)
Comment (caaespin) [reply billing.py]: I think it depends on the context? I personally use `argparse`. However the person that originally wrote the billing portion of this decided to use `click`. So to not disturb the functionality i decided to not move `click` to `argparse`

**Comment by wshands** [comment]
This looks good to me

**Pull request merged**